rfcs/20211118-support-implicit-concat/README.md

```diff
@@ -0,0 +1,132 @@
+  usual in this case. The only thing required is to provide a proper memory
+  handle for a second piece. In this case, everything can be enabled on user
+  side as of current oneDNN state, if using CPU raw pointers or GPU USM memory
+  kind. For OpenCL buffers this trick won't work, unfortunately.
+
+* Scenario 2, when concatenation should happen over the axis which is not the
+  outermost dimension in memory. This is a generic scenario. Same example but
+  for `n > 1` works here. Let's for simplicity take `n = 2`. In this case each
+  primitive which contributes to the final output would have non-trivial strides
+  between `n` points. oneDNN does not provide any API which would allow to user
+  initialize a memory descriptor with both blocking descriptor and strides
+  representing not dense case. But submemory API may be used to create such with
+  two stages: initialize original memory descriptor with tags coming from
+  compute primitive or passed from previous layer, and then perform submemory
+  of this descriptor from the concat destination. This would allow to achieve
+  the desired. This general scenario will utilize offset0 and strides in
+  implementations internally but it requires support from the library and it's
+  not fully under user's control now. But such usage is independent from memory
+  kind.
+
+Also, scenario 2 may become more complicated when adding strided input
+requirement for certain primitives. It may happen in case of branching when
+output of convolution should be put into extended memory and this output would
+be further used as input in a second branch as entry point.
+
+### Implementation
+
+For scenario 2, a simplified algorithm of actions to implement implicit concat
+performance optimization would be the following:
+
+1. Setup primitive descriptors of primitives which would contribute to concat.
+   a. If any of those are compute-intense primitive, querying of destination
+      memory descriptor would be required to get a proper memory format.
+   b. If all of those are memory-bound primitives, query is not required,
+      formats should already be defined and coincide or be compatible.
+2. Setup concat primitive descriptor with destination descriptors from previous
+   step. Omit concat destination to let the library deduce it.
+   a. If everything is good to go, destination would have same blocked format.
+   b. If not, concat would return `abcd`-like format.
+3. Verify that destination was created properly. This can be achieved by
+   creating a memory descriptor with concat destination dims and plain
+   `abcd`-like layout. If they match with queried, concat failed to prepare an
+   implicit setup. Consider using an explicit one.
+4. Create submemory descriptors from concat destination memory descriptor.
+   a. If both memory descriptors were created properly, it means that implicit
+      concat is good to go.
+   b. It may happen that not best optimized implementation is taken, but such
```

Comment:
"that not best...is taken" ---> "that a suboptimal implementation is chosen"

Reply:
Thanks for a suggestion, applied.